traffic_ops/traffic_ops_golang/cdni/shared.go

```diff
@@ -68,17 +82,388 @@ func GetCapabilities(w http.ResponseWriter, r *http.Request) {
+				if err != nil {
+					api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("updating capacity: %w", err))
+					return
+				}
+
+				if rowsAffected, err := result.RowsAffected(); err != nil {
+					api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("updating capacity: getting rows affected: %w", err))
+					return
+				} else if rowsAffected < 1 {
+					api.HandleErr(w, r, inf.Tx.Tx, http.StatusNotFound, fmt.Errorf("no capacity found for update: host: %s, type: %s, limit: %v", host, updatedData.Type, capLim), nil)
+					return
+				} else if rowsAffected > 1 {
+					api.HandleErr(w, r, inf.Tx.Tx, http.StatusInternalServerError, nil, fmt.Errorf("capacity update affected too many rows: %d", rowsAffected))
+					return
+				}
+			}
+		}
+	}
+
+	if asyncErr := api.UpdateAsyncStatus(db, api.AsyncSucceeded, "Capacity requested update has been completed.", asyncId, true); asyncErr != nil {
+		log.Errorf("updating async status for id %v: %v", asyncId, asyncErr)
+	}
+	status, err := deleteCapabilityRequest(reqId, inf.Tx.Tx)
+	if err != nil {
+		api.HandleErr(w, r, inf.Tx.Tx, status, nil, fmt.Errorf("deleting capacity request from queue: %w", err))
+		return
+	}
+	api.WriteResp(w, r, "Successfully updated configuration.")
+}
+
+func getCapabilityIdFromFootprints(updatedData CapacityLimit, ucdn string, inf *api.APIInfo) (int, error) {
+	var potentialCapabilityIds []int
+	for _, footprint := range updatedData.Footprints {
+		selectQuery := `SELECT footprint_type, footprint_value, capability_id FROM cdni_footprints WHERE ucdn = $1 AND footprint_type = $2 AND footprint_value = `
+		valuesAsQuery := "ARRAY['" + strings.Join(footprint.FootprintValue, "','") + "']"
+		rows, err := inf.Tx.Tx.Query(selectQuery+valuesAsQuery, ucdn, footprint.FootprintType)
+		if err != nil {
+			return 0, fmt.Errorf("querying for capacity update request: %w", err)
+		}
+		defer rows.Close()
+		var capabilityIds []int
+		rowCount := 0
+		for rows.Next() {
+			var capabilityId int
+			var footprint Footprint
+			if err := rows.Scan(&footprint.FootprintType, pq.Array(&footprint.FootprintValue), &capabilityId); err != nil {
+				return 0, fmt.Errorf("scanning db rows: %w", err)
+			}
+			rowCount++
+			capabilityIds = append(capabilityIds, capabilityId)
+		}
+
+		if rowCount == 0 {
+			return 0, fmt.Errorf("no capabilities found for footprint: %v, %v", footprint.FootprintType, footprint.FootprintValue)
+		}
+		var newPotentials []int
+		for _, capId := range capabilityIds {
+			if len(potentialCapabilityIds) == 0 {
+				newPotentials = append(newPotentials, capId)
+			} else {
+				if util.IntInArray(potentialCapabilityIds, capId) {
+					newPotentials = append(newPotentials, capId)
+				}
+			}
+		}
+		if len(newPotentials) == 0 {
+			return 0, fmt.Errorf("no capabilities found that match all footprints: %v", updatedData.Footprints)
+		}
+		potentialCapabilityIds = newPotentials
```

Comment:
I'm a little confused by this code. Can you explain what it's doing/checking in terms of capabilities and footprints?

Reply:
yea its super confusing and also im very open to better ways to do this if you have any. 

for this, the footprints can overlap and can be linked to multiple capabilities in the long run. right now this is just for Capacity but there will be a lot more as i keep going. plus each capability can have multiple footprints. so for this code im looping through the footprints that have the correct uCDN, type, and value based on the request, then making a list of potential capability ids based on all of the footprints that came linked to the info in the request.  

this way if you send a request linked to footprints like `[ {footprint-type: ipv4cidr", "footprint-value": ["127.0.0.1", "127.0.0.2" ] }, {footprint-type: countrycode", "footprint-value": ["us"] } ]` then it wont get capabilities linke to just the countrycode footprint or just the ipv4cidr footprint or other capabilities linked to those plus more.

does that make any sense?

Reply:
It might make more sense if I knew the spec better, but don't really want to hold up your PR for that. As long as you think it's per the spec that's good enough for me.

Reply:
revamped it to work way better with some ugly sql